can/interfaces/socketcan/socketcan.py

```diff
@@ -418,92 +437,157 @@ def capture_message(sock):
-        if timeout:
-            try:
-                # get all sockets that are ready (can be a list with a single value
-                # being self.socket or an empty list if self.socket is not ready)
-                ready_receive_sockets, _, _ = select.select([self.socket], [], [], timeout)
-            except OSError:
-                # something bad happened (e.g. the interface went down)
-                log.exception("Error while waiting for timeout")
-                ready_receive_sockets = False
-        else:
-            ready_receive_sockets = True
+        # get all sockets that are ready (can be a list with a single value
+        # being self.socket or an empty list if self.socket is not ready)
+        try:
+            # get all sockets that are ready (can be a list with a single value
+            # being self.socket or an empty list if self.socket is not ready)
+            ready_receive_sockets, _, _ = select.select([self.socket], [], [], timeout)
+        except socket.error as exc:
+            # something bad happened (e.g. the interface went down)
+            raise can.CanError("Failed to receive: %s" % exc)
 
         if ready_receive_sockets: # not empty or True
-            return capture_message(self.socket), self._is_filtered
+            get_channel = self.channel == ""
+            msg = capture_message(self.socket, get_channel)
+            if not msg.channel and self.channel:
+                # Default to our own channel
+                msg.channel = self.channel
+            return msg, self._is_filtered
         else:
             # socket wasn't readable or timeout occurred
             return None, self._is_filtered
 
     def send(self, msg, timeout=None):
+        """Transmit a message to the CAN bus.
+
+        :param can.Message msg: A message object.
+        :param float timeout:
+            Wait up to this many seconds for the transmit queue to be ready.
+            If not given, the call may fail immediately.
+
+        :raises can.CanError:
+            if the message could not be written.
+        """
         log.debug("We've been asked to write a message to the bus")
         logger_tx = log.getChild("tx")
         logger_tx.debug("sending: %s", msg)
 
-        if timeout:
-            # Wait for write availability
-            _, ready_send_sockets, _ = select.select([], [self.socket], [], timeout)
-            if not ready_send_sockets:
-                raise can.CanError("Timeout while sending")
+        started = time.time()
+        # If no timeout is given, poll for availability
+        if timeout is None:
+            timeout = 0
+        time_left = timeout
+        data = build_can_frame(msg)
 
+        while time_left >= 0:
+            # Wait for write availability
+            ready = select.select([], [self.socket], [], time_left)[1]
+            if not ready:
+                # Timeout
+                break
+            sent = self._send_once(data, msg.channel)
```

Comment:
This is a nice solution! But maybe it would be more efficient to use `socket.sendall()` when there is native support?

Reply:
The problem is that `sendall` doesn't have a timeout argument. And it is probably very unusual, if even impossible, that a message could be only partially cued. For most cases it will succeed on first try, or it will fail completely on first `select`.